### PR TITLE
feat: add OAuth token exchange proxy for secure client-secret management

### PR DESCRIPTION
**Pull request by altendky**
## Summary

- **OAuth token exchange proxy** (`workers/oauth-proxy/`): Cloudflare Worker that holds the `client_secret` server-side, providing `/token/exchange` and `/token/refresh` endpoints behind IP allowlisting with DNS-over-HTTPS hostname resolution. Sans-IO architecture with handler.ts (pure logic) + index.ts (thin I/O executor).

- **Rust server proxy support**: New `proxy_url` field on `OAuthTokenData` (mutually exclusive with `client_secret`), `RefreshMethod::Proxy` variant, and `try_refresh_proxy()` with automatic IPv4 retry when the proxy reports a 403 with an IPv6 `source_ip`.

- **OpenCode auth plugin proxy mode**: "Onshape OAuth (via proxy)" method that fetches `client_id` from the proxy's `/config`, exchanges tokens via the proxy, and handles dual-stack IPv4/IPv6 retry by resolving the proxy hostname to all AAAA/A records and connecting to each IP directly via `node:https` (works around Bun ignoring `localAddress` and `family: 4`).

- **Auth status fix**: `Expired` status can now be overridden by `Valid` validation, since the I/O layer proactively refreshes expired tokens before executing the validation request.

- **Proxy always returns JSON**: Upstream non-JSON responses (HTML error pages, empty bodies) are wrapped in a JSON error envelope.

## Key design decisions

- Proxy holds the client secret so CLI users never need it — only `proxy_url` is stored in token files
- `ALLOWED_SOURCES` supports both IP addresses and hostnames (resolved per-request via DNS-over-HTTPS)
- 403 responses include `source_ip` for client-side IPv4 retry and user debugging
- Both Rust and TypeScript clients automatically retry with the other address family on 403 with IPv6 source
- OpenCode plugin does a preflight IP check before sending the user to the browser for OAuth

## Testing

- 35 vitest tests for the proxy worker (all endpoints and edge cases)
- 161-line bun:test suite for the TypeScript IPv4 retry parsing/decision logic
- Rust unit tests for `build_ipv4_retry_client`, proxy URL serde, `from_raw`, proxy auth resolution, and expired-token validation override

### CODE REVIEW

File: npm/opencode-auth/src/ipv4-retry.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  url: string,
+  jsonBody: unknown,
+): Promise<HttpResult | null> {
+  for (const addr of addresses) {
+    try {
+      return await httpsPostJsonToIp(addr, url, jsonBody);
+    } catch {
+      // Connection error — try the next address.
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (altendky) [reply npm/opencode-auth/src/ipv4-retry.ts]: Rather than error-code classification (which is fragile — missing any code means a legitimate connection failure bubbles up as unhandled), restructured `tryAddresses` to parse the URL and serialize the JSON body once before the loop. Non-network errors now propagate immediately to the caller without being silently caught.

[`bf79a9c`](https://github.com/altendky/onshape-mcp/commit/bf79a9c5c41073b69f0703533d2716412ba96b94)

File: npm/opencode-auth/src/ipv4-retry.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        res.on("end", () => {
+          resolve({ status: res.statusCode ?? 0, body: data });
+        });
+      },
+    );
+
+    req.on("error", reject);
+    req.write(body);
+    req.end();
+  });
```
[the remark this answers is not in the document]
Comment (altendky) [reply npm/opencode-auth/src/ipv4-retry.ts]: Added `req.setTimeout(30_000, ...)` with `req.destroy` — using 30s to match the Rust implementation's `reqwest` timeout for the same proxy requests. Also added `res.on("error", reject)` to handle response stream errors that would otherwise leave the promise unsettled.

[`8ed1b95`](https://github.com/altendky/onshape-mcp/commit/8ed1b951419d30a521a601b6504eb3727f6188d3)

File: npm/opencode-auth/src/plugin.ts [hunk 3]
```diff
@@ -39,32 +48,402 @@ function tokenFilePath(): string {
...
+      const url = new URL(req.url);
+      if (url.pathname === "/callback") {
+        const error = url.searchParams.get("error");
+
+        if (error) {
+          (server as any)._callbackUrl = url;
+          return new Response(
+            `<html><body><h1>Authorization Failed</h1><p>${error}</p><p>You can close this window.</p></body></html>`,
+            { headers: { "Content-Type": "text/html" } },
+          );
```
[the remark this answers is not in the document]
Comment (altendky) [reply npm/opencode-auth/src/plugin.ts]: Added an `escapeHtml` helper that escapes the five HTML-special characters (`&`, `<`, `>`, `"`, `'`). Low risk given the ephemeral localhost server, but trivial fix for defense-in-depth.

[`7d5a019`](https://github.com/altendky/onshape-mcp/commit/7d5a019b919526b3d715435f80d2d6e547a074d4)

File: npm/opencode-auth/src/plugin.ts [hunk 4]
```diff
@@ -39,32 +48,402 @@ function tokenFilePath(): string {
...
+                  }
+
+                  const result = JSON.parse(respBody!) as {
+                    access_token: string;
+                    refresh_token?: string;
+                    token_type?: string;
+                    expires_in?: number;
+                    scope?: string;
+                  };
+
```
[the remark this answers is not in the document]
Comment (altendky) [reply npm/opencode-auth/src/plugin.ts]: Added a runtime check that `access_token` is a non-empty string after `JSON.parse`, matching what `oauth4webapi` does for the direct mode path. This brings proxy mode to parity with direct mode's validation rigor and prevents silently storing empty tokens that would cause confusing 401 errors downstream.

[`f4879c5`](https://github.com/altendky/onshape-mcp/commit/f4879c5424bdb2087f800731e09bca011c7cea86)

File: workers/oauth-proxy/src/handler.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  if (value.includes(":")) return true;
+  // IPv4: digits and dots, at least one dot
+  if (/^\d{1,3}(\.\d{1,3}){3}$/.test(value)) return true;
+  return false;
+}
+
+/** Check whether a source IP is in the resolved allowed-IP list. */
+export function isIpAllowed(sourceIp: string, allowedIps: string[]): boolean {
+  return allowedIps.includes(sourceIp);
+}
```
[the remark this answers is not in the document]
Comment (altendky) [reply workers/oauth-proxy/src/handler.ts]: Added `ipaddr.js` as a runtime dependency and updated `isIpAllowed` to parse and normalize both sides for comparison. A fast-path string comparison avoids the parsing overhead when textual forms already agree (which covers the common case where Cloudflare uses canonical form). Added tests covering equivalent IPv6 representations.

[`ff77601`](https://github.com/altendky/onshape-mcp/commit/ff77601ccefc54234e9481b5a7c0205653deec17)

File: npm/opencode-auth/src/plugin.ts [hunk 6]
```diff
@@ -39,32 +48,420 @@ function tokenFilePath(): string {
...
+            if (!configResp.ok) {
+              throw new Error(
+                `Failed to fetch proxy config: ${configResp.status} ${await configResp.text()}`,
+              );
+            }
+            const proxyConfig = (await configResp.json()) as {
+              client_id: string;
+            };
+            const clientId = proxyConfig.client_id;
+
```
[the remark this answers is not in the document]
Comment (altendky) [reply npm/opencode-auth/src/plugin.ts]: Added runtime validation in [`834fbf7`](https://github.com/altendky/onshape-mcp/commit/834fbf7811931ea68a4714e33d43e086ad31bca2). Now checks that `client_id` is a non-empty string before proceeding, matching the existing `access_token` validation pattern from `f4879c5`. Without this, a misconfigured proxy would send the user to Onshape with `client_id=undefined` and produce an opaque error page.
